### ISSUE TITLE
kubelet high cpu usage

### ISSUE THREAD

**Issue by jfchevrette** [kubernetes/kubernetes#18515]
I'm experiencing high CPU usage from the kubelet in my test environment

- minion is running on a single-core VM
- Kubernetes 1.1.2
- single minion, running 40 pods
- 1 container per pod, running apache and being idle
- 1 service, mapping 2 ports from each pod to 2 nodeports (kube-proxy running userspace mode)
- both the kubelet and docker daemon are consuming CPU consistently. stopping the kubelet makes the docker daemon stop consuming CPU

![kubelet-docker-cpu](https://cloud.githubusercontent.com/assets/3001/11721586/84ccf2da-9f32-11e5-9bbc-0c23094b4951.png)

My guess is that the kubelet is gathering stats from docker. Running strace against the kubelet shows that it's constantly reading from /sys/fs/cgroups/

```strace -c -f```
```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 86.67    0.948175         747      1269       210 futex
  6.58    0.072013         381       189           epoll_wait
  5.43    0.059423           9      6264           select
  1.23    0.013447           2      7761         1 read
  0.04    0.000399           0      1220       423 stat
  0.03    0.000367           0      3735           open
  0.01    0.000083           0      3786           close
  0.01    0.000083           0      2800           fstat
  0.01    0.000078           0     16413           clock_gettime
  0.00    0.000000           0        19           write
```

Running ```go tool pprof``` on ```http://localhost:8001/api/v1/proxy/nodes/MINION/debug/pprof/profile```did not seem to work. The profile is pretty much empty and typing 'web' at the (pprof) prompt generated an empty graph. I'm new to this maybe I did not run it properly.

![pprof](https://cloud.githubusercontent.com/assets/3001/11721679/045a9f84-9f33-11e5-8dab-69a3a417fbc2.png)

Is this expected when running so "few" pods/containers?

Please let me know if there are anything else I can provide to help locate the source of this high CPU usage.

**Comment by hanikesn** [comment]
What docker version are you using? Docker v.1.9.x has know performance problems:
https://github.com/docker/docker/issues/17720

**Comment by jfchevrette** [comment]
I am running docker 1.9.1. 

I do not seem to be running into the issue described in that issue. docker ps/run/stats all executes well under 1 sec

**Comment by yujuhong** [comment]
@jfchevrette, your screenshot showed ~24% of cpu usage from kubelet at 40 pods. This doesn't seem particularly high to me. I measured the cpu usage of kubelet/docker for v1.1 in #16943. Do you see higher usage than that?

We understand the usage is high and we are actively working on the issue for v1.2. The latest update is in https://github.com/kubernetes/kubernetes/issues/16943#issuecomment-162606927. 

You might also want to know that a goroutine leak in cadvisor (built in as part of kubelet) has been fixed (#16651) and has been included in the release v1.1.3. The goroutine leak may cause increase in cpu usage over time. 


> Running go tool pprof on http://localhost:8001/api/v1/proxy/nodes/MINION/debug/pprof/profiledid not seem to work.

My guess is that you didn't pass the kubelet binary to the command. It should be
`go tool pprof <path_to_kubelet_binary> http://localhost:8001/api/v1/proxy/nodes/MINION/debug/pprof/profile`

**Comment by yujuhong** [comment]
@jfchevrette, I am closing this issue due to inactivity. Please follow #16943 for the progress of reducing kubelet/docker cpu usage for v1.2. Also, we've found issues with docker 1.9 that have kept us from upgrading. We are planning to support docker v1.10 directly.

Feel free to reopen the issue if you have any more information.
